server/s3_post_complete_action.cc

```diff
@@ -424,77 +440,152 @@ bool S3PostCompleteAction::validate_parts() {
+              bucket_metadata->get_extended_metadata_index_oid());
+
+      object_metadata->set_extended_object_metadata(new_ext_object_metadata);
+      s3_log(S3_LOG_DEBUG, stripped_request_id,
+             "Created extended object metadata to store object parts");
+    }
+  }
+  const std::shared_ptr<S3ObjectExtendedMetadata>& new_object_ext_metadata =
+      object_metadata->get_extended_object_metadata();
+
+  if (new_object_ext_metadata) {
+    // Add object part to new object's extended metadata
+    struct s3_part_frag_context part_frag_ctx = {};
+    part_frag_ctx.motr_OID = part_metadata->get_oid();
+    part_frag_ctx.PVID = part_metadata->get_pvid();
+    part_frag_ctx.versionID = object_metadata->get_obj_version_key();
+    part_frag_ctx.item_size = part_metadata->get_content_length();
+    part_frag_ctx.layout_id = part_metadata->get_layout_id();
+    part_frag_ctx.is_multipart = true;
+    const int part_number =
+        strtoul(part_metadata->get_part_number().c_str(), NULL, 0);
+    // For multipart object, pass the part_no as object part number
+    // and fragment_no=1 (part is not further fragmented)
+    new_object_ext_metadata->add_extended_entry(part_frag_ctx, 1, part_number);
+    object_metadata->set_number_of_fragments(
+        new_object_ext_metadata->get_fragment_count());
+    object_metadata->set_number_of_parts(
+        new_object_ext_metadata->get_part_count());
+    s3_log(S3_LOG_DEBUG, request_id,
+           "Added object part [%d] with object oid"
+           "[%" SCNx64 " : %" SCNx64 "] to extended md\n",
+           part_number, part_frag_ctx.motr_OID.u_hi,
+           part_frag_ctx.motr_OID.u_lo);
+  }
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+// Add object parts of specified object (old or new) to
+// the specified probable dead oid list
+void S3PostCompleteAction::add_part_object_to_probable_dead_oid_list(
+    const std::shared_ptr<S3ObjectMetadata>& object_metadata,
+    std::vector<std::unique_ptr<S3ProbableDeleteRecord>>&
+        parts_probable_del_rec_list,
+    bool is_old_object) {
   s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  const std::shared_ptr<S3ObjectExtendedMetadata>& object_ext_metadata =
+      object_metadata->get_extended_object_metadata();
+  unsigned int total_parts = 0;
+  if (object_ext_metadata &&
+      ((total_parts = object_ext_metadata->get_part_count()) > 0)) {
+    s3_log(S3_LOG_DEBUG, request_id, "Total parts available [%u]\n",
+           total_parts);
+    // object has some parts
+    const std::map<int, std::vector<struct s3_part_frag_context>>&
+        ext_entries_mp = object_ext_metadata->get_raw_extended_entries();
+    struct m0_uint128 old_oid, part_list_index_oid;
+    bool is_multipart = false;
+    for (unsigned int key_indx = 1; key_indx <= total_parts; key_indx++) {
+      const std::vector<struct s3_part_frag_context>& part_entry =
+          ext_entries_mp.at(key_indx);
```

Comment:
std::map<...>::at(...) can throw exceptions.
Can we be sure that this won't happen?

Reply:
It can happen when 'total_parts' (which is the value for field "PARTS" in object list index) is more than actual number of entries in the extended md index for the same object with version. This is less likely to occur, until the entries are removed externally from extended md index for the object/version OR, there is some error in saving parts to part list index table during multipart upload api, which is less likely to occur because we return service unavailable or internal error, when there is any error in writing part entry to part index in multipart upload.

So, less likely to happen, but can occur. Will try to guard it using exception.